File: ros_ws/src/robot_control/robot_control/kinematics_publisher.py
#!/usr/bin/python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import Float64MultiArray
from math import pi, sqrt, atan2

class KinematicsPublisher(Node):
    
    def __init__(self):
        super().__init__("kinematics_publisher")
        # subscriber
        self.sub_vel = self.create_subscription(Twist, "cmd_vel", self.cmdvel_callback, 10)
        self.sub_vel
        # publisher
        self.pub_pos = self.create_publisher(Float64MultiArray, '/forward_position_controller/commands', 10)
        self.pub_vel = self.create_publisher(Float64MultiArray, '/forward_velocity_controller/commands', 10)
        # timer
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        # varieble
        self.wheel_vel = [0.0, 0.0, 0.0, 0.0]
        self.swerve_pos = [0.0, 0.0, 0.0, 0.0]
        
    def timer_callback(self):   
        vel_msg = Float64MultiArray()
        pos_msg = Float64MultiArray()
        vel_msg.data = self.wheel_vel
        pos_msg.data = self.swerve_pos 
        self.pub_vel.publish(vel_msg)
        self.pub_pos.publish(pos_msg)

    def cmdvel_callback(self, msg):

        Vx = 0.0
        Vy = 0.0
        W = 0.0
        theta = 0.0
        
        Vy = msg.linear.y
        Vx = msg.linear.x
        W = msg.angular.z
        
        V = sqrt(Vx**2 + Vy**2) 
        
        if Vy == 0:
            if Vx > 0:
                theta = pi/2
            elif Vx < 0:
                theta = -pi/2
        else:
            theta = atan2(Vx,Vy)
                   
        self.wheel_vel = [V, V, V, V]
        self.swerve_pos = [theta, theta, theta, theta]
        

def main(args=None):
    rclpy.init(args=args)

    kinematics_publisher = KinematicsPublisher()

    rclpy.spin(kinematics_publisher)

    kinematics_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros_ws/src/robot_control_cpp/src/kinematics_publisher.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <vector>

#include "kinematics_calculation.hpp"

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float64_multi_array.hpp>
#include <tf2_msgs/msg/tf_message.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <geometry_msgs/msg/twist.hpp>

using std::placeholders::_1;
using namespace std::chrono_literals;

class KinematicsPublisher : public rclcpp::Node
{

public:
	KinematicsPublisher() : Node("kinematics_publisher"), Vx_(0.0), Vy_(0.0), W_(0.0)
	{
		pub_pos = this->create_publisher<std_msgs::msg::Float64MultiArray>("/forward_position_controller/commands", 10);
		pub_vel = this->create_publisher<std_msgs::msg::Float64MultiArray>("/forward_velocity_controller/commands", 10);
		sub_cmd_vel = this->create_subscription<geometry_msgs::msg::Twist>("cmd_vel", 10, std::bind(&KinematicsPublisher::cmdvel_callback, this, _1));
		sub_tf = this->create_subscription<tf2_msgs::msg::TFMessage>("/tf", 10, std::bind(&KinematicsPublisher::tf_callback, this, _1));
		timer_ = this->create_wall_timer(500ms, std::bind(&KinematicsPublisher::timer_callback, this));
	}

private:
	void timer_callback()
	{
		auto vel_msg = std_msgs::msg::Float64MultiArray();
		auto pos_msg = std_msgs::msg::Float64MultiArray();
		vel_msg.data = wheel_vel_;
		pos_msg.data = swerve_pos_;
		pub_vel->publish(vel_msg);
		pub_pos->publish(pos_msg);
	}

	void tf_callback(const tf2_msgs::msg::TFMessage::SharedPtr msg)
	{

		std::vector<std::string> target_child_frame_id = {"swerve_fl_link",
														  "swerve_fr_link",
														  "swerve_rl_link",
														  "swerve_rr_link"};

		for (const auto &transform : msg->transforms)
		{
			for (int child_frame_id_check = 0; child_frame_id_check < 4; ++child_frame_id_check)
			{
				if (transform.child_frame_id == target_child_frame_id[child_frame_id_check])
				{
					tf_[child_frame_id_check][0] = transform.transform.translation.x;
					tf_[child_frame_id_check][1] = transform.transform.translation.y;
				}
			}
		}
	}

	void cmdvel_callback(const geometry_msgs::msg::Twist::SharedPtr msg)
	{
		Vx_ = msg->linear.x;
		Vy_ = msg->linear.y;
		W_ = msg->angular.z;

		kinematics.kinematics_calculation(Vx_, Vy_, W_, tf_);

		wheel_vel_ = kinematics.get_wheel_vel();
		swerve_pos_ = kinematics.get_swerve_pos();
	}

	rclcpp::TimerBase::SharedPtr timer_;
	rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr pub_pos;
	rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr pub_vel;
	rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_cmd_vel;
	rclcpp::Subscription<tf2_msgs::msg::TFMessage>::SharedPtr sub_tf;
	size_t count_;
	float_t Vx_, Vy_, W_;
	std::vector<double> wheel_vel_ = {0.0, 0.0, 0.0, 0.0};
	std::vector<double> swerve_pos_ = {0.0, 0.0, 0.0, 0.0};
	std::vector<std::vector<double>> tf_ = {{0.0, 0.0}, {0.0, 0.0}, {0.0, 0.0}, {0.0, 0.0}};
	KinematicsCalculation kinematics;
};

int main(int argc, char *argv[])
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<KinematicsPublisher>());
	rclcpp::shutdown();
	return 0;
}